server.go

```diff
@@ -1886,7 +1888,9 @@ func (s *Server) ShutdownWithContext(ctx context.Context) (err error) {
 	}
 
 	if s.done != nil {
-		close(s.done)
+		s.closeDone.Do(func() {
+			close(s.done)
+		})
```

Comment:
You have to reset `s.closeDone` when `s.done` is reset. This change now makes it impossible to reuse the Server struct after closing it.

Reply:
oh I see, and since you want to reuse the Server struct, you may create the s.done again

I must think about this, because if I reset s.closeDone we may ended with the same code and same behaviour

Reply:
I don't mean here exactly. You have to reset it when a Serve is called I think. That way you don't have the same behaviour anymore, only when you reuse the Server struct and then it's intended.

Reply:
I think it can be modified like this
`func (ctx *RequestCtx) Done() <-chan struct{} {
       //fix  Use locks to prevent concurrent modifications, and use new variables to prevent panic caused by modifying the original done chan to nil.
	ctx.s.mu.Lock()
	defer ctx.s.mu.Unlock()
	if ctx.s.done == nil {
		tmp := make(chan struct{}, 1)
		tmp <- struct{}{}
		return tmp
	}
	doneChan := ctx.s.done
	return doneChan
}
func (ctx *RequestCtx) Err() error {
	select {
	case <-ctx.Done():    //fix  Use unified functions instead of reference variables to converge fetching into one place
		return context.Canceled
	default:
		return nil
	}
}
`

Reply:
I think it can be modified like this
```
func (ctx *RequestCtx) Done() <-chan struct{} {
	//fix  Use locks to prevent concurrent modifications, 
	//and use new variables to prevent panic caused by modifying the original done chan to nil.
	ctx.s.mu.Lock()
	defer ctx.s.mu.Unlock()
	if ctx.s.done == nil {
		tmp := make(chan struct{}, 1)
		tmp <- struct{}{}
		return tmp
	}
	doneChan := ctx.s.done
	return doneChan
}
func (ctx *RequestCtx) Err() error {
	select {
	case <-ctx.Done(): // //fix  Use unified functions instead of reference variables to converge fetching into one place
		return context.Canceled
	default:
		return nil
	}
}
```